src/com/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -928,119 +757,245 @@
   "Given a certname, querys the DB for existing facts for that
    certname and will update, delete or insert the facts as necessary
-   to match the facts argument."
-  [{:keys [name values environment timestamp producer-timestamp] :as fact-data} :- facts-schema]
-  (let [factset (certname-to-factset-id name)
-        old-facts (current-fact-value-ids factset)
-        new-facts (new-fact-value-ids values)]
-    (sql/update-values :factsets ["id=?" factset]
-                       {:timestamp (to-timestamp timestamp)
-                        :environment_id (ensure-environment environment)
-                        :producer_timestamp (to-timestamp producer-timestamp)})
-    (utils/diff-fn (zipmap new-facts (repeat nil))
-                   (zipmap old-facts (repeat nil))
-                   #(insert-facts! factset %)
-                   #(delete-facts! factset %)
-                   identity)))
+   to match the facts argument. (cf. add-facts!)"
+  [{:keys [name values environment timestamp producer-timestamp] :as fact-data}
+   :- facts-schema]
+
+  (sql/transaction
+   (let [factset-id (certname-to-factset-id name)
+         initial-factset-paths-vhashes
+         (query-to-vec "SELECT fp.path, fv.value_hash FROM facts f
+                          INNER JOIN fact_paths fp ON f.fact_path_id = fp.id
+                          INNER JOIN fact_values fv ON f.fact_value_id = fv.id
+                          WHERE factset_id = ?"
+                       factset-id)
+         ;; Ensure that all the required paths and values exist.
+         paths-and-valuemaps (facts/facts->paths-and-valuemaps values)
+         pathstrs (map (comp facts/factpath-to-string first) paths-and-valuemaps)
+         valuemaps (map second paths-and-valuemaps)
+         vhashes (map :value_hash valuemaps)
+         paths-to-ids (realize-paths! pathstrs)
+         vhashes-to-ids (realize-values! valuemaps)
+         ;; Add new facts and remove obsolete facts.
+         replacement-pv-pairs (set (map #(vector (paths-to-ids %1)
+                                                 (vhashes-to-ids %2))
+                                        pathstrs vhashes))
+         current-pairs (set (select-pid-vid-pairs-for-factset factset-id))
+         [new-pairs rm-pairs] (data/diff replacement-pv-pairs current-pairs)]
+
+     ;; Paths are unique per factset so we can delete solely based on pid.
+     (when rm-pairs
+       (let [rm-pids (set (map first rm-pairs))]
+         (sql/do-prepared
+          (format "DELETE FROM facts WHERE factset_id = ? AND fact_path_id %s"
+                  (jdbc/in-clause rm-pids))
+          (cons factset-id rm-pids))))
+
+     (insert-facts-pv-pairs! factset-id new-pairs)
+
+     (when rm-pairs
+       (delete-pending-path-id-orphans! factset-id
+                                        (set/difference
+                                         (set (map first current-pairs))
+                                         (set (map first replacement-pv-pairs))))
```

Comment:
I haven't tested, but is this expressible in terms of rm-pairs? (set (map first rm-pairs)) or something maybe? it seems like we might be able to avoid diffing both here and above somehow.